### ISSUE TITLE
CombinedCharts:Highlight particular bar from stack bar tap

### ISSUE THREAD

**Issue by senthilmgs** [PhilJay/MPAndroidChart#1302]
I tried a sample to show linechart, barcharts(stacked bar) with negative values by using combinedchart view from mpandroidcharts lib. I'm facing an issue to higjlight the selected tap onto to barcharts +ve/-ve stack value.It's hard to identify the -ve stack value. It always hghligths the +ve stack value.Suggest me to highlight the particular stack bar of user preferred either-ve/+ve stack.I just need to highlight the particular tap selection on to this chart view. Attached a screenshot for your reference here.
![device-2015-12-22-190211](https://cloud.githubusercontent.com/assets/1173989/11971823/0a82f6b4-a969-11e5-9b45-310fe434ffc2.png)

**Comment by ardmn** [comment]
Hello. This occurs because BarChart use BarHighlighter and CombinedChart use CombinedHighlighter. CombinedHighlighter override only getSelectionDetailsAtIndex method of super class ChartHighlighter so it not support _Stacked_ and _Grouped_ BarChar data sets. 

I have not good solution for this problem but I have workaround that works for me(and my needs).

First I show source code of sample Activity and MyCombinedChartHighlighter next  :

```java
package com.xxmassdeveloper.mpchartexample;

import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.app.FragmentActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.WindowManager;

import com.github.mikephil.charting.charts.CombinedChart;
import com.github.mikephil.charting.charts.CombinedChart.DrawOrder;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.XAxis.XAxisPosition;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.CombinedData;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.interfaces.datasets.IDataSet;

import java.util.ArrayList;

public class CombinedChartActivityExample extends FragmentActivity {

    private CombinedChart mChart;

    protected String[] mMonths = new String[] {
            "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_combined_example);

        mChart = (CombinedChart) findViewById(R.id.chart1);
        mChart.setHighlighter(new MyCombinedChartHighlighter(mChart));
        mChart.setDrawGridBackground(false);
        mChart.setDescription("");

        // scaling can now only be done on x- and y-axis separately
        mChart.setPinchZoom(false);

        mChart.setDrawBarShadow(false);
        mChart.setDrawValueAboveBar(true);

        mChart.getAxisRight().setEnabled(false);
        mChart.getAxisLeft().setAxisMaxValue(25f);
        mChart.getAxisLeft().setAxisMinValue(-25f);
        mChart.getAxisLeft().setDrawGridLines(true);
        mChart.getAxisLeft().setDrawZeroLine(true);
        mChart.getAxisLeft().setLabelCount(7, false);
        mChart.getAxisLeft().setTextSize(9f);

        XAxis xAxis = mChart.getXAxis();
        xAxis.setPosition(XAxisPosition.BOTH_SIDED);
        xAxis.setDrawGridLines(false);
        xAxis.setDrawAxisLine(false);
        xAxis.setTextSize(9f);

        Legend l = mChart.getLegend();
        l.setPosition(Legend.LegendPosition.BELOW_CHART_LEFT);
        l.setFormSize(8f);
        l.setFormToTextSpace(4f);
        l.setXEntrySpace(6f);

        // draw bars behind lines
        mChart.setDrawOrder(new DrawOrder[]{
                DrawOrder.BAR, DrawOrder.BUBBLE, DrawOrder.CANDLE, DrawOrder.LINE, DrawOrder.SCATTER
        });


        CombinedData data = new CombinedData(mMonths);

        data.setData(generateLineData());
        data.setData(generateBarData());


        mChart.setData(data);
        mChart.invalidate();
    }

    private LineData generateLineData() {

        LineData d = new LineData();

        ArrayList<Entry> entries = new ArrayList<Entry>();

        for (int index = 0; index < 12; index++)
            entries.add(new Entry((int) (getRandom(10, 15) * (index % 3 == 0 ? -1 : 1)), index));

        LineDataSet set = new LineDataSet(entries, "Line DataSet");
        set.setColor(Color.rgb(240, 238, 70));
        set.setLineWidth(1.5f);
        //set.setCircleColor(Color.rgb(240, 238, 70));
        //set.setCircleRadius(5f);
        set.setFillColor(Color.rgb(240, 238, 70));
        set.setDrawCubic(true);
        set.setDrawValues(true);
        set.setValueTextSize(10f);
        set.setValueTextColor(Color.rgb(0, 238, 70));

        set.setAxisDependency(YAxis.AxisDependency.LEFT);

        d.addDataSet(set);

        return d;
    }

    private BarData generateBarData() {

        // IMPORTANT: When using negative values in stacked bars, always make sure the negative values are in the array first
        ArrayList<BarEntry> yValues = new ArrayList<BarEntry>();
        yValues.add(new BarEntry(new float[]{-10, 10}, 0));
        yValues.add(new BarEntry(new float[]{-12, 13}, 1));
        yValues.add(new BarEntry(new float[]{-15, 15}, 2));
        yValues.add(new BarEntry(new float[]{-17, 17}, 3));
        yValues.add(new BarEntry(new float[]{-19, 20}, 4));
        yValues.add(new BarEntry(new float[]{-19, 19}, 5));
        yValues.add(new BarEntry(new float[]{-16, 16}, 6));
        yValues.add(new BarEntry(new float[]{-13, 14}, 7));
        yValues.add(new BarEntry(new float[]{-10, 11}, 8));
        yValues.add(new BarEntry(new float[]{-5, 6}, 9));
        yValues.add(new BarEntry(new float[]{-1, 2}, 10));

        BarDataSet set = new BarDataSet(yValues, "Age Distribution");
        set.setValueTextSize(7f);
        set.setAxisDependency(YAxis.AxisDependency.LEFT);
        set.setBarSpacePercent(40f);
        set.setColors(new int[]{Color.BLUE, Color.RED});
        set.setStackLabels(new String[]{
                "Men", "Women"
        });

        String[] xVals = new String[]{"0-10", "10-20", "20-30", "30-40", "40-50", "50-60", "60-70", "70-80", "80-90", "90-100", "100+"};

        BarData data = new BarData(xVals, set);

        return data;
    }


    private float getRandom(float range, float startsfrom) {
        return (float) (Math.random() * range) + startsfrom;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.combined, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.actionToggleLineValues: {
                for (IDataSet set : mChart.getData().getDataSets()) {
                    if (set instanceof LineDataSet)
                        set.setDrawValues(!set.isDrawValuesEnabled());
                }

                mChart.invalidate();
                break;
            }
            case R.id.actionToggleBarValues: {
                for (IDataSet set : mChart.getData().getDataSets()) {
                    if (set instanceof BarDataSet)
                        set.setDrawValues(!set.isDrawValuesEnabled());
                }

                mChart.invalidate();
                break;
            }
        }
        return true;
    }

}
```

```java
package com.xxmassdeveloper.mpchartexample;

import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.ChartData;
import com.github.mikephil.charting.data.CombinedData;
import com.github.mikephil.charting.highlight.ChartHighlighter;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.highlight.Range;
import com.github.mikephil.charting.interfaces.dataprovider.BarLineScatterCandleBubbleDataProvider;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.github.mikephil.charting.interfaces.datasets.IDataSet;
import com.github.mikephil.charting.utils.SelectionDetail;

import java.util.List;

public class MyCombinedChartHighlighter extends ChartHighlighter<BarLineScatterCandleBubbleDataProvider> {

    public MyCombinedChartHighlighter(BarLineScatterCandleBubbleDataProvider chart) {
        super(chart);
    }

    private BarData getBarData()
    {
        BarData barData = null;

        CombinedData data = (CombinedData) mChart.getData();
        List<ChartData> dataObjects = data.getAllData();
        for(ChartData chartData: dataObjects)
            if(chartData instanceof BarData)
            {
                barData = (BarData)chartData;
                break;
            }

        return barData;
    }

    private int getDataIndex()
    {
        CombinedData data = (CombinedData) mChart.getData();
        List<ChartData> dataObjects = data.getAllData();

        for(int i =0 ; i< dataObjects.size();i++)
            if(dataObjects.get(i) instanceof BarData)
               return i;

        return 0;
    }

    @Override
    public Highlight getHighlight(float x, float y) {

        BarData barData = getBarData();

        final int xIndex = getXIndex(x);
        final float baseNoSpace = getBase(x);
        final int setCount = barData.getDataSetCount();
        int dataSetIndex = ((int)baseNoSpace) % setCount;

        if (dataSetIndex < 0) {
            dataSetIndex = 0;
        } else if (dataSetIndex >= setCount) {
            dataSetIndex = setCount - 1;
        }

        SelectionDetail selectionDetail = getSelectionDetail(xIndex, y, dataSetIndex);
        if (selectionDetail == null)
            return null;

        IBarDataSet set = barData.getDataSetByIndex(dataSetIndex);
        if (set.isStacked()) {

            float[] pts = new float[2];
            pts[1] = y;

            // take any transformer to determine the x-axis value
            mChart.getTransformer(set.getAxisDependency()).pixelsToValue(pts);

            return getStackedHighlight(selectionDetail,
                    set,
                    xIndex,
                    pts[1]);
        }

        return new Highlight(
                xIndex,
                selectionDetail.value,
                getDataIndex(),//selectionDetail.dataIndex,
                selectionDetail.dataSetIndex,
                -1);
    }

    @Override
    protected int getXIndex(float x) {

        if (!getBarData().isGrouped()) {
            return super.getXIndex(x);
        } else {

            float baseNoSpace = getBase(x);

            int setCount = getBarData().getDataSetCount();
            int xIndex = (int) baseNoSpace / setCount;

            int valCount = mChart.getData().getXValCount();

            if (xIndex < 0)
                xIndex = 0;
            else if (xIndex >= valCount)
                xIndex = valCount - 1;

            return xIndex;
        }
    }

    @Override
    protected SelectionDetail getSelectionDetail(int xIndex, float y, int dataSetIndex) {

        dataSetIndex = Math.max(dataSetIndex, 0);

        BarData barData = getBarData();
        IDataSet dataSet = barData.getDataSetCount() > dataSetIndex
                ? barData.getDataSetByIndex(dataSetIndex)
                : null;
        if (dataSet == null)
            return null;

        final float yValue = dataSet.getYValForXIndex(xIndex);

        if (yValue == Double.NaN) return null;

        return new SelectionDetail(
                yValue,
                dataSetIndex,
                dataSet);
    }

    /**
     * This method creates the Highlight object that also indicates which value of a stacked BarEntry has been selected.
     *
     * @param selectionDetail the selection detail to work with looking for stacked values
     * @param set
     * @param xIndex
     * @param yValue
     * @return
     */
    protected Highlight getStackedHighlight(
            SelectionDetail selectionDetail,
            IBarDataSet set,
            int xIndex,
            double yValue) {

        BarEntry entry = set.getEntryForXIndex(xIndex);

        if (entry == null)
            return null;

        if (entry.getVals() == null) {
            return new Highlight(xIndex,
                    entry.getVal(),
                    getDataIndex(),//selectionDetail.dataIndex,
                    selectionDetail.dataSetIndex);
        }

        Range[] ranges = getRanges(entry);
        if (ranges.length > 0) {
            int stackIndex = getClosestStackIndex(ranges, (float)yValue);
            return new Highlight(
                    xIndex,
                    entry.getPositiveSum() - entry.getNegativeSum(),
                    getDataIndex(),//selectionDetail.dataIndex,
                    selectionDetail.dataSetIndex,
                    stackIndex,
                    ranges[stackIndex]
            );
        }

        return null;
    }

    /**
     * Returns the index of the closest value inside the values array / ranges (stacked barchart) to the value given as
     * a parameter.
     *
     * @param ranges
     * @param value
     * @return
     */
    protected int getClosestStackIndex(Range[] ranges, float value) {

        if (ranges == null || ranges.length == 0)
            return 0;

        int stackIndex = 0;

        for (Range range : ranges) {
            if (range.contains(value))
                return stackIndex;
            else
                stackIndex++;
        }

        int length = Math.max(ranges.length - 1, 0);

        return (value > ranges[length].to) ? length : 0;

    }

    /**
     * Returns the base x-value to the corresponding x-touch value in pixels.
     *
     * @param x
     * @return
     */
    protected float getBase(float x) {

        // create an array of the touch-point
        float[] pts = new float[2];
        pts[0] = x;

        // take any transformer to determine the x-axis value
        mChart.getTransformer(YAxis.AxisDependency.LEFT).pixelsToValue(pts);
        float xVal = pts[0];

        int setCount = getBarData().getDataSetCount();

        // calculate how often the group-space appears
        int steps = (int) ((float) xVal / ((float) setCount + getBarData().getGroupSpace()));

        float groupSpaceSum = getBarData().getGroupSpace() * (float) steps;

        float baseNoSpace = (float) xVal - groupSpaceSum;
        return baseNoSpace;
    }

    /**
     * Splits up the stack-values of the given bar-entry into Range objects.
     *
     * @param entry
     * @return
     */
    protected Range[] getRanges(BarEntry entry) {

        float[] values = entry.getVals();

        if (values == null || values.length == 0)
            return new Range[0];

        Range[] ranges = new Range[values.length];

        float negRemain = -entry.getNegativeSum();
        float posRemain = 0f;

        for (int i = 0; i < ranges.length; i++) {

            float value = values[i];

            if (value < 0) {
                ranges[i] = new Range(negRemain, negRemain + Math.abs(value));
                negRemain += Math.abs(value);
            } else {
                ranges[i] = new Range(posRemain, posRemain + value);
                posRemain += value;
            }
        }

        return ranges;
    }

}

```

What I do ? 
I copied code from the **com.github.mikephil.charting.highlight.BarHighlighter.java** and added two methods:
- private BarData getBarData()
- private int getDataIndex()

I changed code of BarHighlighter  where needed to get the BarData ( replaced  to new method  getBarData() ). And I change everywhere Highlight dataIndex parameter to getDataIndex() because we need get correct data in method:
  
```java
public Entry com.github.mikephil.charting.data.CombinedData.getEntryForHighlight(Highlight highlight) 
{
...
 ChartData data = dataObjects.get(highlight.getDataIndex());
...
}
```

**Comment by PhilJay** [comment]
This issue will be fixed in v3.0.0 which is out in a few weeks.
